File: Lesson 7, Stacks and Queues/Lesson 7, Stacks and Queues, Fish.c
int solution(int A[], int B[], int N)
{
    int alive = N;
    for (int i = 0, j = 0, stack[100000]; i < N; i++)
        if (B[i])
            stack[j++] = i;
        else if (j) {
            alive--;
            if (A[stack[j-1]] < A[i])
                j--, i--;
        }
    return alive;
}
// O(N)

File: Lesson 7, Stacks and Queues/Lesson 7, Stacks and Queues, Fish.java
import java.util.*;
class Solution {
    public int solution(int[] A, int[] B) {
        int i = 0, alive, N; alive = N = A.length;
        for (ArrayDeque<Integer> stack = new ArrayDeque<Integer>(); i < N; i++)
            if (B[i] == 1)
                stack.addLast(i);
            else if (!stack.isEmpty()) {
                alive--;
                if (A[stack.getLast()] < A[i]) {
                    stack.removeLast();
                    i--;
                }
            }
        return alive;
    }
}
// O(N)